app/src/org/commcare/activities/ConnectionDiagnosticActivity.java

```diff
@@ -55,31 +53,32 @@ protected void onCreate(Bundle savedInstanceState) {
-                                        Localization.get("connection.task.internet.fail")
-                                        : Localization.get("connection.task.remote.ping.fail");
-
-                                receiver.txtInteractiveMessages.setText(displayMessage);
-                                receiver.txtInteractiveMessages.setVisibility(View.VISIBLE);
-
-                                receiver.settingsButton.setVisibility(View.VISIBLE);
-                            } else if (failedTest == Test.commCarePing) {
-                                //unable to ping commcare -- report this to cchq
-                                receiver.txtInteractiveMessages.setText(
-                                        Localization.get("connection.task.commcare.html.fail"));
-                                receiver.txtInteractiveMessages.setVisibility(View.VISIBLE);
-
-                                receiver.reportButton.setVisibility(View.VISIBLE);
-                            } else if (failedTest == null) {
-                                receiver.txtInteractiveMessages.setText(Localization.get("connection.task.success"));
-                                receiver.txtInteractiveMessages.setVisibility(View.VISIBLE);
-                                receiver.settingsButton.setVisibility(View.INVISIBLE);
-                                receiver.reportButton.setVisibility(View.INVISIBLE);
+                            switch (failedTest) {
+                                case CONNECTED:
+                                    receiver.txtInteractiveMessages.setText(Localization.get("connection.task.success"));
+                                    receiver.txtInteractiveMessages.setVisibility(View.VISIBLE);
+                                    receiver.settingsButton.setVisibility(View.INVISIBLE);
+                                    receiver.reportButton.setVisibility(View.INVISIBLE);
+                                    break;
+                                case COMMCARE_DOWN:
+                                    //unable to ping commcare -- report this to cchq
+                                    receiver.txtInteractiveMessages.setText(Localization.get("connection.task.commcare.html.fail"));
+                                    receiver.txtInteractiveMessages.setVisibility(View.VISIBLE);
+                                    receiver.reportButton.setVisibility(View.VISIBLE);
+                                    break;
+                                case DISCONNECTED:
+                                case CAPTIVE_PORTAL:
+                                    //get the appropriate display message based on what the problem is
+                                    String displayMessage = failedTest == ConnectivityStatus.NetworkState.DISCONNECTED ?
+                                            Localization.get("connection.task.internet.fail")
+                                            : Localization.get("connection.task.remote.ping.fail");

```

Comment:
Instead of giving a general message like `connection.task.remote.ping.fail` , since we have pretty strong evidence that we are inside a captive portal , we should switch to giving a captive portal specific message to the user here.